[OS-893] Initial code submission and debian packaging

skarbat opened the pull request.

skarbat committed:
Initial code submission and debian packaging

skarbat committed:
Fix dependency name, rectify python modules install path

skarbat commented:
Package is now building fine and installs correctly on a Kano OS 4.2.1 image.

skarbat committed:
Correct wrong path name

radujipa commented on python/qa_tools/hid.py:
```diff
@@ -0,0 +1,37 @@
+qa_hid='sudo /usr/bin/qa-hid'
+
+
+def call_hid(command_line):
```
```suggestion
def _call_hid(command_line):
```
I'd recommend prefixing this function with an underscore to explicitly say this is function is internal to the module.

radujipa commented on python/qa_tools/hid.py:
```diff
@@ -0,0 +1,37 @@
+
+import os
+
+qa_hid='sudo /usr/bin/qa-hid'
```
I'd keep this under the `_call_hid` function.

radujipa commented on tools/qa-findimage/rpi-screenshot.h:
```diff
@@ -0,0 +1,8 @@
+/*
+ *  rpi-screenshot.h
+ *
```
Copyright?

radujipa commented on tools/qa-findimage/rpi-screenshot.cpp:
```diff
@@ -0,0 +1,146 @@
+/*
+ *  rpi-screenshot.cpp
+ *
+ *  Copyright (C) 2016 Kano Computing Ltd.
```
```suggestion
 *  Copyright (C) 2019 Kano Computing Ltd.
```

radujipa commented on tools/qa-findimage/rpi-screenshot.cpp:
```diff
@@ -0,0 +1,146 @@
+ *  rpi-screenshot.cpp
+ *
+ *  Copyright (C) 2016 Kano Computing Ltd.
+ *  License:   http://www.gnu.org/licenses/gpl-2.0.txt GNU General Public License v2
```
```suggestion
 *  License: http://www.gnu.org/licenses/gpl-2.0.txt GNU GPL v2
```

radujipa commented on tools/qa-findimage/qa-findimage.cpp:
```diff
@@ -0,0 +1,196 @@
+/*
+ *  file: findimage.cpp
+ *
+ * Copyright (C) 2016 Kano Computing Ltd.
```
```suggestion
 *  Copyright (C) 2019 Kano Computing Ltd.
```

radujipa commented on tools/qa-findimage/qa-findimage.cpp:
```diff
@@ -0,0 +1,196 @@
+ *  file: findimage.cpp
+ *
+ * Copyright (C) 2016 Kano Computing Ltd.
+ * License:   http://www.gnu.org/licenses/gpl-2.0.txt GNU General Public License v2
```
```suggestion
 * License: http://www.gnu.org/licenses/gpl-2.0.txt GNU GPL v2
```

tombettany commented on tools/qa-findimage/qa-findimage.cpp:
```diff
@@ -0,0 +1,196 @@
+}
+
+
+void MatchingMethod (int, void*)
```
Huh? The parameters aren't used? If not then mark them as unused:

```c++
void MatchingMethod(__attribute__((unused)) int foo, __attribute__((unused)) void *bar)
{
    ...
```

skarbat replied:
fixed

tombettany commented on tools/qa-findimage/rpi-screenshot.cpp:
```diff
@@ -0,0 +1,146 @@
+ *  Conditional compilation is in place to allow for building on other platforms.
+ *
+ */
+
```
```c++
#include "rpi-screenshot.h"
```

skarbat replied:
fixed

tombettany commented on tools/qa-findimage/rpi-screenshot.cpp:
```diff
@@ -0,0 +1,146 @@
+        *pheight = height;
+    }
+
+    return ((void *)dmxImagePtr);
```
This is already a `void *`, why cast? Also, use modern casting `static_cast<void *>(dmxImagePtr)`.

skarbat replied:
redundant cast, removed

tombettany commented on tools/qa-findimage/rpi-screenshot.h:
```diff
@@ -0,0 +1,8 @@
+ *
+ */
+
+void *get_rpi_screenshot(bool verbose,  int *pwidth, int *pheight);
```
Header guards

skarbat replied:
fixed

tombettany commented on tools/qa-hid/hidmain.cpp:
```diff
@@ -0,0 +1,141 @@
+
+  // Parsing command line arguments with docopt.
+  std::map<std::string, docopt::value> args =
+    docopt::docopt(USAGE, {argv + 1, argv + argc});
```
Nice use of docopt!

tombettany commented on tools/qa-hid/hidmain.cpp:
```diff
@@ -0,0 +1,141 @@
+
+int main(int argc, char **argv)
+{
+  int uifd;
```
Switched from 4 spaces for indents to 2 - make this guy 4 spaces.

radujipa replied:
Please use 4 spaces for indents for consistency

tombettany commented on tools/qa-hid/hidlib.cpp:
```diff
@@ -0,0 +1,393 @@
+/* hidlib.cpp */
```
Is this and `hidlib.h` included completely unmodified? If so, include them in a subdirectory called `vendor` with a README about the source. If not, document our changes.

tombettany commented on tools/qa-findimage/qa-findimage.cpp:
```diff
@@ -0,0 +1,196 @@
+void MatchingMethod (int, void*);
+
+// Global Variables
+Mat img; Mat templ; Mat result, result_threshold;
```
One definition per line.

tombettany commented on tools/qa-findimage/qa-findimage.cpp:
```diff
@@ -0,0 +1,196 @@
+string output_image;
+
+// Function Headers
+void MatchingMethod( int, void* );
```
Redefinition?

tombettany commented on tools/qa-findimage/qa-findimage.cpp:
```diff
@@ -0,0 +1,196 @@
+
+// function prototypes
+int main (int argc, char** argv);
+void MatchingMethod (int, void*);
```
Move these into the header file?

tombettany commented on tools/qa-findimage/qa-findimage.cpp:
```diff
@@ -0,0 +1,196 @@
+using namespace cv;
+
+// function prototypes
+int main (int argc, char** argv);
```
Do we really need to pre-define `main()`?

tombettany commented on tools/qa-findimage/qa-findimage.cpp:
```diff
@@ -0,0 +1,196 @@
+#include "rpi-screenshot.h"
+
+using namespace std;
+using namespace cv;
```
Can be somewhat risky `using namespace`, use with caution.

tombettany commented on tools/qa-findimage/qa-findimage.cpp:
```diff
@@ -0,0 +1,196 @@
+
+    // Localizing the best match with minMaxLoc
+    double minVal; double maxVal, accuracy_val;
+    Point minLoc; Point maxLoc;
```
One def per line

tombettany commented on tools/qa-findimage/qa-findimage.cpp:
```diff
@@ -0,0 +1,196 @@
+
+    // the magic name "rpi" for source image means a RaspberryPI screenshot
+    if (!strcmp(CMDLINE_DISPMANX_SOURCE, source_image.c_str())) {
+
```
Weird newline

tombettany commented on tools/qa-findimage/qa-findimage.cpp:
```diff
@@ -0,0 +1,196 @@
+            cout << "error: could not take a dispmanx screenshot" << endl;
+            return(1);
+        }
+        else {
```
`else` linked to previous line:

```c++
} else {
```

also applies elsewhere.

tombettany commented on tools/qa-findimage/qa-findimage.cpp:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    // the magic name "rpi" for source image means a RaspberryPI screenshot
+    if (!strcmp(CMDLINE_DISPMANX_SOURCE, source_image.c_str())) {
```
Probably cleaner to do the opposite:

```c++
if (std::string(CMDLINE_DISPMANX_SOURCE) != source_img) {
```

or might be even better to

```c++
const std::string CMDLINE_DISPMANX_SOURCE("dispmanx");
// ...
// int main()
// ...
    if (CMDLINE_DISPMANX_SOURCE != source_image) {
```

tombettany commented on tools/qa-findimage/qa-findimage.cpp:
```diff
@@ -0,0 +1,196 @@
+
+// Provide this literal as the source image to find image on the RPi screen
+#define CMDLINE_DISPMANX_SOURCE   "dispmanx"
+#define CMDLINE_SHOW_UI           "ui"
```
Likewise here, make this a `const std::string`

tombettany commented on tools/qa-findimage/qa-findimage.cpp:
```diff
@@ -0,0 +1,196 @@
+{
+    string source_image, image_to_find;
+
+    if (argc < 3) {
```
Docopt?

skarbat replied:
Definitely yes, Tom - will address that later, thanks!

tombettany commented on tools/qa-findimage/qa-findimage.cpp:
```diff
@@ -0,0 +1,196 @@
+        return(1);
+    }
+    else {
+        source_image=argv[1];
```
Spaces around `=`

```c++
source_image = argv[1];
```

Likewise everywhere else.

tombettany commented on tools/qa-findimage/qa-findimage.cpp:
```diff
@@ -0,0 +1,196 @@
+
+    if (argc < 3) {
+        cout << "Syntax: findimage <source image> <image to find> [output image]" << endl;
+        return(1);
```
Unnecessary parentheses

tombettany commented on tools/qa-findimage/rpi-screenshot.cpp:
```diff
@@ -0,0 +1,146 @@
+ *  in RGB colorspace.
+ *
+ */
+void *get_rpi_screenshot(bool verbose, int *pwidth, int *pheight)
```
Don't like this just returning `void *`, can we do something like a `typedef` to at least be able to validate the type?

tombettany commented on tools/qa-findimage/rpi-screenshot.cpp:
```diff
@@ -0,0 +1,146 @@
+        vc_dispmanx_display_close(displayHandle);
+
+        printf("vc_dispmanx_rect_set() failed\n");
+	bcm_host_deinit();
```
Tab amongst spaces.

tombettany commented on tools/qa-findimage/rpi-screenshot.cpp:
```diff
@@ -0,0 +1,146 @@
+
+        printf("vc_dispmanx_rect_set() failed\n");
+	bcm_host_deinit();
+        return(NULL);
```
We're using C++, `NULL` should never be used as it is simply an alias to `0`, use `nullptr`

tombettany replied:
Also unnecessary parentheses.

tombettany commented on tools/qa-findimage/rpi-screenshot.cpp:
```diff
@@ -0,0 +1,146 @@
+ *
+ */
+
+#ifndef RPI
```
This might be better to do via a separate file so this guy would just be

```c++
#ifdef RPI
    #include "platforms/rpi/rpi-screenshot.cpp"
#else
    #include "platforms/generic/rpi-screenshot.cpp"
#endif
```

and the implementations would be in each of those included files.

tombettany commented on tools/qa-findimage/qa-findimage.cpp:
```diff
@@ -0,0 +1,196 @@
+        return(1);
+    }
+
+    MatchingMethod (0, 0);
```
No space between the method name and the `(`. Likewise everywhere.

tombettany commented on python/qa_tools/hid.py:
```diff
@@ -0,0 +1,37 @@
+
+
+def key_press(key):
+    command_line='{} key-press {}'.format(qa_hid, key)
```
Spacing around `=`

tombettany commented on python/qa_tools/hid.py:
```diff
@@ -0,0 +1,37 @@
+
+
+
+    
```
Way too many newlines here.

tombettany commented on python/qa_tools/hid.py:
```diff
@@ -0,0 +1,37 @@
+#
```
Might be worth running this through `pylint`

tombettany reviewed (approved):
Great start. Comments are mainly for readability and maintainability.

radujipa commented on tools/qa-hid/hidlib.h:
```diff
@@ -0,0 +1,32 @@
+
```
Same copyright as the cpp?

radujipa commented on tools/qa-hid/hidmain.cpp:
```diff
@@ -0,0 +1,141 @@
+/*  
+ *  fakemain.c
+ *  
```
copyright

radujipa reviewed (changes requested):
Few bits and bobs

skarbat committed:
specify unused parameters

skarbat committed:
missing include file

skarbat committed:
redundant cast

skarbat committed:
header duplication guards

tombettany commented on tools/qa-findimage/rpi-screenshot.h:
```diff
@@ -0,0 +1,12 @@
+ *
+ */
+
+#ifndef RPI_SCREENSHOT
```
```suggestion
#ifndef RPI_SCREENSHOT
#define RPI_SCREENSHOT
```

lauraenria committed:
bug: suggestion hid.py

tombettany committed:
Lint: Add cpplint configuration

Add a default `cpplint` configuration file.

tombettany committed:
Style: Fix linting errors

Modify styling to remove linting errors.

tombettany committed:
Git Ignore: Add gitignore file

Add basic `gitignore` file.

tombettany committed:
Platform: Add stub for `bcm_host.h`

Stub out functions provided by `bcm_host.h` to enable compilation when
not on RPi.

Also stub out Linux headers.

tombettany committed:
C++11 Style: Switch to use smart pointers

Convert the `malloc()` usage to utilise `std::shared_ptr`s to prevent
memory leaks.

tombettany committed:
Constants: Switch from `#define` to `const char *`

When using `#define` for constants, no type checking is done so prefer
typed variables marked as `const` and let the compiler do its thing.

tombettany committed:
Makefile: Clean-up `CFLAGS` and `LDFLAGS`

Simplify the `Makefile` to utilise standard `CFLAGS` and `LDFLAGS`
variables

tombettany committed:
Style: Fix linting of python scripts

Resolve linting errors from the Python scripts.

tombettany commented:
@skarbat A couple of outstanding questions about copyright and licensing which we should resolve before merging in.

skarbat committed:
Implementation for "key-press KEY_TAB" option

skarbat committed:
Fixed incorrect key code for the ENTER key

skarbat committed:
No need to define key codes, linux provides it for us

skarbat committed:
Fix for an image buffer getting out of scope, and source image was lost

skarbat committed:
Removed redundant blank lines

skarbat committed:
Implementation for absolute mouse movement

skarbat committed:
Adding tests for qa-findimage using exact size images

skarbat committed:
Adding tests for scaled up icons - currently failing as expected

skarbat committed:
Adding images for testing full qa-findimage resolution coverage

skarbat committed:
Implement scale up/down to locate differing sized assets - all tests passed

skarbat committed:
Remove "time" tool, pytest already reports this info

skarbat committed:
Add docopt to qa-findimage

skarbat committed:
qa-hid mouse-move tests

skarbat committed:
remoe obsolete test module

skarbat committed:
Add qa-hid tests through the python interface module

skarbat committed:
Disable qa-hid tests to build the package (momentarily)

skarbat committed:
Test and python implementation for mouse move with scale

skarbat committed:
Fix findimage problem comparing 2 screenshots - test included
